hydromt/models/components/mesh.py

```diff
@@ -10,34 +10,389 @@
+        )
+
+        if crs:  # Restore crs
+            for grid in self.data.ugrid.grids:
+                grid.set_crs(crs)
+
+        # # update related geoms if necessary: region TODO: Check if is still needed
+        # if overwrite_grid or new_grid:
+        #     # add / updates region
+        #     if "region" in self.geoms:
+        #         self._geoms.pop("region", None)
+        #     _ = self.region
+
+    def write(
+        self,
+        fn: str = "mesh/mesh.nc",
+        write_optional_ugrid_attributes: bool = True,
+        **kwargs,
+    ) -> None:
+        """Write model grid data to a netCDF file at <root>/<fn>.
+
+        Keyword arguments are passed to :py:meth:`xarray.Dataset.to_netcdf`.
+
+        Parameters
+        ----------
+        fn : str, optional
+            Filename relative to the model root directory, by default 'grid/grid.nc'.
+        write_optional_ugrid_attributes : bool, optional
+            If True, write optional ugrid attributes to the netCDF file, by default
+            True.
+        **kwargs : dict
+            Additional keyword arguments to be passed to the
+            `xarray.Dataset.to_netcdf` method.
+        """
+        if self.data is None:
+            self.logger.debug("No mesh data found, skip writing.")
+            return
+        if not self.model_root.is_writing_mode():
+            raise IOError("Model opened in read-only mode")
+        # filename
+        _fn = join(self.model_root.path, fn)
+        if not isdir(dirname(_fn)):
+            os.makedirs(dirname(_fn))
+        self.logger.debug(f"Writing file {fn}")
+        ds_out = self.data.ugrid.to_dataset(
+            optional_attributes=write_optional_ugrid_attributes,
+        )
+        if self.crs is not None:
+            # save crs to spatial_ref coordinate
+            ds_out = ds_out.rio.write_crs(self.crs)
+        ds_out.to_netcdf(_fn, **kwargs)
+
+    def read(
+        self, fn: str = "mesh/mesh.nc", crs: Optional[Union[CRS, int]] = None, **kwargs
+    ) -> None:
+        """Read model mesh data at <root>/<fn> and add to mesh property.
+
+        key-word arguments are passed to :py:meth:`~hydromt.models.Model.read_nc`
+
+        Parameters
+        ----------
+        fn : str, optional
+            filename relative to model root, by default 'mesh/mesh.nc'
+        crs : CRS or int, optional
+            Coordinate Reference System (CRS) object or EPSG code representing the
+            spatial reference system of the mesh file. Only used if the CRS is not
+            found when reading the mesh file.
+        **kwargs : dict
+            Additional keyword arguments to be passed to the `read_nc` method.
+        """
+        if not self.model_root.is_reading_mode():
+            raise IOError("Model not opend in read mode")
+        ds = xr.merge(
+            read_nc(
+                fn, root=self.model_root.path, logger=self.logger, **kwargs
```

Comment:
```suggestion
                fn, root=self._model_root.path, logger=self._logger, **kwargs
```

Do you have tests for these, or should you update from the refactor-models branch?